README.md

```diff
@@ -217,9 +217,23 @@ Benchee.run(%{
 
 ```
 
-### Setup and teardown
+### Hooks (Setup, Teardown etc.)
 
-If you want to do setup and teardown, i.e. do something before or after the benchmarking suite executes this is very easy in benchee. As benchee is just plain old functions just do it before/after you call benchee:
+If you want to do something before or after something happens in benchee, we got you coverd! In general, there are 3 distrinct times where you might want to do something before/after it happened:
+
+* the whole suite (aka before everything/after everything) - done by just running code before/after benchee
+* an individual benchmarking scenario (one result reported you, combination of a benchmarking function and possibly an input) - `before_scenario`/`after_scenario`
+* each individual invocation of a benchmarking function - `before_each`/`after_each`
+
+benchee covers all these cases (well the first one doesn't need any specific support) and each of them will be discussed here followed by a little overview when what actually happens.
+
+Of course, **hooks are not included in the measured run times**. So they are there espacially if you want to do something and want it to **not be included in the measured times**. Sadly there is the notable exception of _too_fast_functions_ (the ones that execute in less than 10 microseconds). As we need to measure their repeated invocations to get halfway good measurements `before_each` and `after_each` hooks are included there
+
+Before you dig deep into these, **most of the times you shouldn't actually need these**. When you are just benchmarking plain old immutable functions no hooks whatsoever should be needed. Hooks are mostly needed when you need side effects to happen, randomize something or whatever.
+
```

Comment:
I think this belongs at the beginning, to help people orient themselves. Most users can skip this section :)

Reply:
True that!